src/Microsoft.SqlTools.ServiceLayer/Profiler/ProfilerService.cs

```diff
@@ -1,420 +1,505 @@
+                }
+            });
+        }
+
+        private async Task StartLocalFileSession(StartProfilingParams parameters, RequestContext<StartProfilingResult> requestContext)
         {
-            monitor.StopMonitoringSession(parameters.OwnerUri, out _);
+            var xeSession = XEventSessionFactory.OpenLocalFileSession(parameters.SessionName);
+            monitor.StartMonitoringSession(parameters.OwnerUri, xeSession);
+            xeSession.Start();
+            var result = new StartProfilingResult() { UniqueSessionId = xeSession.Id.ToString(), CanPause = false };
+            await requestContext.SendResult(result);
         }
 
-        /// <summary>
-        /// Gets a list of all running XEvent Sessions
-        /// </summary>
-        /// <returns>
-        /// A list of the names of all running XEvent sessions
-        /// </returns>
-        internal List<string> GetXEventSessionList(string ownerUri, ConnectionInfo connInfo)
-        {
-            var sqlConnection = ConnectionService.OpenSqlConnection(connInfo);
-            SqlStoreConnection connection = new SqlStoreConnection(sqlConnection);
-            BaseXEStore store = CreateXEventStore(connInfo, connection);
-
-            // get session names from the session list
-            List<string> results = store.Sessions.Aggregate(new List<string>(), (result, next) =>
+        /// <summary>
+        /// Handle request to stop a profiling session
+        /// </summary>
+        internal async Task HandleStopProfilingRequest(StopProfilingParams parameters, RequestContext<StopProfilingResult> requestContext)
+        {
+            await Task.Run(async () =>
             {
-                result.Add(next.Name);
-                return result;
+                try
+                {
+                    monitor.StopMonitoringSession(parameters.OwnerUri, out ProfilerSession session);
+
+                    if (session != null)
+                    {
+                        // Occasionally we might see the InvalidOperationException due to a read is 
+                        // in progress, add the following retry logic will solve the problem.
+                        int remainingAttempts = 3;
+                        while (true)
+                        {
+                            try
+                            {
+                                session.XEventSession.Stop();
+                                session.Dispose();
+                                await requestContext.SendResult(new StopProfilingResult { });
+                                break;
+                            }
+                            catch (InvalidOperationException)
+                            {
+                                remainingAttempts--;
+                                if (remainingAttempts == 0)
+                                {
+                                    throw;
+                                }
+                                Thread.Sleep(500);
```

Comment:
> Thread.Sleep(500); [](http://example.com/codeflow?start=31&length=19)

maybe the Sleeps should be replaced with `await Task.Delay(500)` now

Reply:
Fixed.